lib/active_model_serializers/fragment_cache.rb

```diff
@@ -0,0 +1,107 @@
+module ActiveModelSerializers
+  class FragmentCache
+    attr_reader :serializer
+
+    def initialize(adapter, serializer, options)
+      @instance_options = options
+      @adapter    = adapter
+      @serializer = serializer
+    end
+
+    # TODO: Use Serializable::Resource
+    # TODO: call +constantize+ less
+    # 1. Create a CachedSerializer and NonCachedSerializer from the serializer class
+    # 2. Serialize the above two with the given adapter
+    # 3. Pass their serializations to the adapter +::fragment_cache+
+    def fetch
+      klass = serializer.class
+      # It will split the serializer into two, one that will be cached and one that will not
+      serializers = fragment_serializer(serializer.object.class.name, klass)
+
+      # Instantiate both serializers
+      cached_serializer     = serializers[:cached].constantize.new(serializer.object)
+      non_cached_serializer = serializers[:non_cached].constantize.new(serializer.object)
+
+      cached_adapter     = adapter.class.new(cached_serializer, instance_options)
+      non_cached_adapter = adapter.class.new(non_cached_serializer, instance_options)
+
+      # Get serializable hash from both
+      cached_hash     = cached_adapter.serializable_hash
+      non_cached_hash = non_cached_adapter.serializable_hash
+
+      # Merge both results
+      adapter.fragment_cache(cached_hash, non_cached_hash)
```

Comment:
This is an adapter method since the adapter might want to customize how the two hashes are merged, e.g. see JSON API adapter.


```diff
-          non_cached_hash.merge cached_hash
+          root = false if instance_options.include?(:include)
+          core_cached       = cached_hash.first
+          core_non_cached   = non_cached_hash.first
+          no_root_cache     = cached_hash.delete_if { |key, _value| key == core_cached[0] }
+          no_root_non_cache = non_cached_hash.delete_if { |key, _value| key == core_non_cached[0] }
+          cached_resource   = (core_cached[1]) ? core_cached[1].deep_merge(core_non_cached[1]) : core_non_cached[1]
+          hash = (root) ? { root => cached_resource } : cached_resource
+
+          hash.deep_merge no_root_non_cache.deep_merge no_root_cache
```